### PR TITLE
Fix unfail mechanism based of determinism

### PR DESCRIPTION
**Pull request by otaviopace**
Before the "PoI for failed subgraphs" feature, where we advance the
deployment head a block if the error is deterministic, we would do
the `unfail` logic after a block is successfully processed, like this:

```rust
match process_block() {
  Ok(needs_restart) => {
    if first_run {
      store.unfail();
    }
    // ...
  }
  // ...
}
```

With this feature the `unfail` mechanism had to change, we did that by
changing two things about it:

- It would run before `process_block`
- Making it revert to the parent block if the error was deterministic

This worked fine for deterministic cases, however for non-deterministic
ones we faced an issue.

We need to run `unfail` after `process_block` went successful (returned
`Ok`). This is necessary because we can only unfail the deployment if
the subgraph actually advanced past the error block range.

Example with the `unfail` previous to this commit:

- Subgraph state:
  - failed
  - deployment head at 1399
  - non-determinstic error happened at 1400
- Once the index-node gets restarted, `unfail` would run, but nothing
  would happen since the deployment head only advances after a
  successful `process_block` execution.
- Subgraph would continue to advance it's pointer but `unfail` wouldn't
  run anymore (only once at start). This would make the subgraph be in
  the `failed` state, but advancing it's pointer normally.

To fix the issue, we need to run `unfail` for non-deterministic errors
**after** the `process_block` went successful:

- Subgraph state:
  - failed
  - deployment head at 1399
  - non-determinstic error happened at 1400
- Once the index-node gets restarted, `unfail_determinstic_error` would
  run and be NOOP, then the next block would be processed (1400), the
  error can be fixed (non-deterministic) when `process_block` gets to
  run and returns `Ok`, then `unfail_non_determinstic_error` executes
  and **unfails** the subgraph finally 🙌
- Subgraph continues to advance it's pointer and the status is correct

### CODE REVIEW

File: core/src/subgraph/instance_manager.rs [hunk 1]
```diff
@@ -592,30 +593,33 @@ where
...
+            if should_try_unfail_deterministic {
+                should_try_unfail_deterministic = false;
+
+                if let Some(current_ptr) = ctx.inputs.store.block_ptr()? {
+                    if let Some(parent_ptr) =
+                        ctx.inputs.triggers_adapter.parent_ptr(&current_ptr).await?
+                    {
+                        // This reverts the deployment head to the parent_ptr if
+                        // deterministic errors happened.
+                        // There's no point in calling it if we have no current or parent block
```
Comment (leoyvens) [inline core/src/subgraph/instance_manager.rs:610]: The claim that there is no point in doing this if there are no current or parent ptrs is true but deserves a justification.

File: store/postgres/src/deployment_store.rs [hunk 2]
```diff
@@ -1170,29 +1170,20 @@ impl DeploymentStore {
...
-
+    // Base for unfail logic, basically it early returns / does nothing if:
+    //
+    // - There's no fatal error for a subgraph
+    // - The error's determinism isn't the same as the one requested
+    //
+    // Otherwise it calls the provided closure/function with the subgraph error.
+    fn unfail_error<F>(&self, site: &Arc<Site>, deterministic: bool, f: F) -> Result<(), StoreError>
+    where
+        F: Fn(&PgConnection, &DeploymentHash, &ErrorDetail) -> Result<(), StoreError>,
```
Comment (leoyvens) [inline store/postgres/src/deployment_store.rs:1181]: I dislike this callback style, I'd rather duplicate the function for the deterministic and non-deterministic cases, and then factor out code that can be reused between them.

**Review by leoyvens — CHANGES_REQUESTED** [review]

File: store/postgres/src/deployment_store.rs (hunk 2 above)
Comment (otaviopace) [reply store/postgres/src/deployment_store.rs:1181]: Do you have a suggestion for doing this without the callback style considering that we need early returns here?

I could do a common method, but that would have to return: `should_early_return: bool` + `maybe_subgraph_error: Option<ErrorDetail>` values, which I dislike more over the callback style.

Example:

```rust
fn unfail_error() -> (bool, Option<ErrorDetail>) {

}

fn unfail_deterministic() {
    let (should_early_return, maybe_subgraph_error) = unfail_error();

    if should_early_return {
        return Ok(());
    }

    let subgraph_error = maybe_subgraph_error.unwrap();

    // ... main logic
}

fn unfail_non_deterministic() {
    let (should_early_return, maybe_subgraph_error) = unfail_error();

    if should_early_return {
        return Ok(());
    }

    let subgraph_error = maybe_subgraph_error.unwrap();

    // ... main logic
}
```

File: core/src/subgraph/instance_manager.rs (hunk 1 above)
Comment (otaviopace) [reply core/src/subgraph/instance_manager.rs:610]: Done! 🙂 

Please tell me if it's still not clear enough.

File: store/postgres/src/deployment_store.rs (hunk 2 above)
Comment (otaviopace) [reply store/postgres/src/deployment_store.rs:1181]: IMO when we're talking about having to control flow like this (early returning), the callback style is more suitable.

And in the future if these two methods don't have anything in common anymore, the developer coding over this should remove the common function.

Comment (otaviopace) [reply store/postgres/src/deployment_store.rs]: Abstraction removed, and `error::detail` one improved.

Fixed.

**Review by leoyvens — APPROVED** [review]
Nice! Please squash when merging.

**Pull request merged**